core/src/main/java/io/aiven/klaw/service/utils/CacheService.java

```diff
@@ -0,0 +1,161 @@
+package io.aiven.klaw.service.utils;
+
+import io.aiven.klaw.error.KlawException;
+import io.aiven.klaw.model.ApiResponse;
+import io.aiven.klaw.service.HighAvailabilityUtilsService;
+import java.util.*;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.*;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+/**
+ * This cache provides a standardised way with built in inter instance update to other caches when
+ * working in a multi cluster deployment. Standard querying mechanisms are provided to ensure that
+ * the most efficient use of the cache is used along with the ability to use Java Streams with the
+ * output for complex queries.
+ *
+ * @param <T> The Entity Type that is to be stored in this Cache.
+ */
+@Slf4j
+public class CacheService<T> {
+
+  public static final String URL_SEPERATOR = "/";
+  public static final String TENANT = "tenant";
+  public static final String ID = "id";
+  Map<Integer, Map<Integer, T>> cache;
+
+  RestTemplate rest;
+
+  private final String urlEndpoint;
+
+  private final HighAvailabilityUtilsService utilsService;
+
+  private List<String> clusterUrls;
+
+  public CacheService(String endpoint, HighAvailabilityUtilsService utilsService) {
+    // TODO an interface needs to be added to allow the passing in of different communication
+    // methods (kafka/https/rabbitmq etc) to allow any org to use what they want for maintaining
+    // cache.
+    this.cache = new HashMap<>();
+    this.urlEndpoint = "cache" + URL_SEPERATOR + endpoint;
+    this.utilsService = utilsService;
+  }
+
+  public T add(int tenantId, Integer id, T entry) {
+    return addOrUpdate(tenantId, id, entry, false);
+  }
+
+  public void addAll(int tenantId, Map<Integer, T> entries) {
+    if (entries == null) {
+      entries = new HashMap<>();
+    }
+    getCache(tenantId).putAll(entries);
+  }
+
+  public T remove(int tenantId, Integer id, boolean isLocalUpdate) {
```

Comment:
```suggestion
  public T removeEnv(int tenantId, Integer id, boolean isLocalUpdate) {
```

As this method is applicable only for environment.

Or can we make it generic ?

Reply:
Currently it is only applicable for environment but the plan is to expand the cache to allow us not to require a cache reset when we add a cluster/team/user etc so leaving it as remove as it will be used by all of them